Fix #568: recognize provided expressions

Jand42 opened the pull request:
Expression API recognizes many calls generated by type providers that previously thrown an exception.

Changes included:
* Add a case `TyparRigidity.Unresolved` to notate type parameters with unknown type constraints. These are ignored when doing signature comparison checks
* An unscoping for looking up IL members, version information can be missing from the reference coming from a TP
* Better recursive stripping of type aliases and measures
* Added various tests, TestTP is now included in main solution

One TP test method `ChoiceConstructionAndMatch` fails during F# type checking, union generated subclass fails to resolve to F# symbol, this test is currently commented out

Jand42 commented:
Still not correct for use with FSharp.Data+WebSharper 4, adding more tests

dsyme commented:
@Jand42 Wow, impressive

I need to look very closely at the changes in tast.fs and TastOps.fs - if there are any changes there that you think are unnecessary then taking them out would simplify accepting this

One compilation problem:

    [00:06:05] C:\projects\fsharp-compiler-service\src\fsharp\ConstraintSolver.fs(348,11): error FS0025: Incomplete pattern matches on this expression. For example, the value '(_,Unresolved)' may indicate a case not covered by the pattern(s).

Jand42 commented:
@dsyme Probably all changes in tast.fs and TastOps.fs can be reverted. I am adding more tests to gather and print all method signatures, not just the shape of the expression, I have already found some mistakes.

Jand42 commented:
@dsyme I have reverted every change to tast.fs and TastOps.fs, they were not needed.

Added a printer for method reference signatures, and tests with it.

One limitation remains, that if overloads are created with using the same name in `CompiledName`, it cannot be resolved.

dsyme commented:
Oh that/s great, makes it much easier. 

There's still a failure in CI build

```
[00:08:34] C:\Program Files\dotnet\dotnet.exe compile-fsc @C:\projects\fsharp-compiler-service\tests\service\obj\Release\netcoreapp1.0\dotnet-compile.rsp returned Exit Code 1
[00:08:34] 
[00:08:34] C:\projects\fsharp-compiler-service\tests\service\ExprTests.fs(863,9): error FS0039: The namespace or module 'ProjectCracker' is not defined
[00:08:34] 
```

dsyme commented on src/fsharp/vs/Exprs.fs:
```diff
@@ -752,6 +679,190 @@ module FSharpExprConvert =
+                None
+
+        // Otherwise try to bind it to an F# symbol
+        match try1 with
```
Could you extract this large expression out to a separate helper function please, outside the big ``let rec``? Along the lines of ``bindILMethodRefToSymbol`` in the previous code. Thanks

dsyme commented on src/fsharp/vs/Exprs.fs:
```diff
@@ -752,6 +679,190 @@ module FSharpExprConvert =
+        let makeCall minfo =
+            ConvObjectModelCallLinear cenv env (isNewObj, minfo, enclTypeArgs, methTypeArgs, callArgs) id   
+
+        let makeFSCall isMember (vr: ValRef) =
```
Please extract ``makeFSCall`` and ``makeFSExpr`` out to helper functions in a module at the top of the file, independent of the big ``let rec``.  Along the lines of ``bindILMethodRefToSymbol`` in the previous code.  I'd like to see that they are `///` comments and be independent of the rest of the code.

dsyme commented:
Still seeing the following fail for .NET Core: 

    [00:08:16] C:\projects\fsharp-compiler-service\tests\service\ExprTests.fs(865,9): error FS0039: The namespace or module 'ProjectCracker' is not defined

Jand42 commented:
I have disabled runnint TP tests on .NET core for now.

Added some tests and fixes for accessing module members and `get_Tag` of non-nullable unions.

I have looked into separating whole `ConvILCall` or just the inner helper functions, but it would just make it more convoluted with I think. The difference is that old implementation only returned an `FSharpMemberOrFunctionOrValue` and now it constructs different kinds of expressions, so it can't be on the top of the file without returning a union result which had to be matched to create the final expression. Also, `ConvLValueExpr` is used depending on the kind of F# member matched, otherwise an extra AddressOf would remain in the expression.

dsyme commented:
OK. Great work.  Do you think this is now ready to merge?

Jand42 commented:
I will do some testing with it today on FSharp.Data+WebSharper 4, will report back.

Jand42 commented:
Everything worked ok (building full WebSharper 4 stack, testing try.websharper), I hope to release all very soon

dsyme commented:
6.0.2 released. Awesome job fixing this!!!
